File: Cleanism/Stores/ChoreStore.swift
import Foundation
import SwiftUI

@MainActor
class ChoreStore: ObservableObject {
    @Published var chores: [Chore] = []

    private let saveKey = "cleanism_chores"

    init() {
        load()
    }

    // MARK: - Persistence

    private func load() {
        guard let data = UserDefaults.standard.data(forKey: saveKey),
              let decoded = try? JSONDecoder().decode([Chore].self, from: data) else { return }
        chores = decoded
    }

    private func save() {
        guard let data = try? JSONEncoder().encode(chores) else { return }
        UserDefaults.standard.set(data, forKey: saveKey)
    }

    // MARK: - CRUD

    @discardableResult
    func addChore(
        name: String,
        beforeImagePath: String? = nil,
        contextDescription: String,
        aiTip: String,
        areaTag: AreaTag,
        dueDate: Date? = nil,
        priority: Priority? = nil,
        estimatedMinutes: Int? = nil
    ) -> UUID {
        let chore = Chore(
            name: name,
            beforeImagePath: beforeImagePath,
            contextDescription: contextDescription,
            aiTip: aiTip,
            areaTag: areaTag,
            dueDate: dueDate,
            priority: priority,
            estimatedMinutes: estimatedMinutes
        )
        chores.insert(chore, at: 0)
        save()
        return chore.id
    }

    func updateChore(id: UUID, updates: (inout Chore) -> Void) {
        guard let index = chores.firstIndex(where: { $0.id == id }) else { return }
        updates(&chores[index])
        save()
    }

    func deleteChore(id: UUID) {
        chores.removeAll { $0.id == id }
        save()
    }

    func completeChore(id: UUID) {
        guard let index = chores.firstIndex(where: { $0.id == id }) else { return }
        chores[index].isCompleted = true
        chores[index].completedAt = Date()
        save()
    }

    func uncompleteChore(id: UUID) {
        guard let index = chores.firstIndex(where: { $0.id == id }) else { return }
        chores[index].isCompleted = false
        chores[index].completedAt = nil
        save()
    }

    func addNote(id: UUID, note: String) {
        guard let index = chores.firstIndex(where: { $0.id == id }) else { return }
        chores[index].notes = note
        save()
    }

    // MARK: - Queries

    func choreById(_ id: UUID) -> Chore? {
        chores.first { $0.id == id }
    }

    var pendingChores: [Chore] {
        chores.filter { !$0.isCompleted }
    }

    var completedChores: [Chore] {
        chores.filter { $0.isCompleted }
    }

    func choresByArea(_ area: AreaTag) -> [Chore] {
        chores.filter { $0.areaTag == area && !$0.isCompleted }
    }

    var overdueChores: [Chore] {
        let now = Date()
        return chores.filter { !$0.isCompleted && $0.dueDate != nil && $0.dueDate! < now }
    }

    // MARK: - Streak

    var streak: Int {
        let completed = completedChores
            .compactMap { $0.completedAt }
            .map { Calendar.current.startOfDay(for: $0) }
            .sorted(by: >)

        guard !completed.isEmpty else { return 0 }

        let uniqueDates = Array(Set(completed)).sorted(by: >)
        let today = Calendar.current.startOfDay(for: Date())
        var count = 0

        for i in 0..<uniqueDates.count {
            let expected = Calendar.current.date(byAdding: .day, value: -i, to: today)!
            if uniqueDates.contains(expected) {
                count += 1
            } else if i == 0, let first = uniqueDates.first,
                      first == Calendar.current.date(byAdding: .day, value: -1, to: today) {
                count += 1
            } else {
                break
            }
        }

        return count
    }
}

File: cleanism backup/mobile/src/state/choreStore.ts
import { create } from "zustand";
import { createJSONStorage, persist } from "zustand/middleware";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { Chore, AreaTag, Priority } from "../types/chore";
import { v4 as uuidv4 } from "uuid";

interface ChoreState {
  chores: Chore[];
}

interface ChoreActions {
  addChore: (chore: Omit<Chore, "id" | "createdAt" | "isCompleted">) => string;
  updateChore: (id: string, updates: Partial<Chore>) => void;
  deleteChore: (id: string) => void;
  completeChore: (id: string) => void;
  uncompleteChore: (id: string) => void;
  addNote: (id: string, note: string) => void;
  getChoreById: (id: string) => Chore | undefined;
  getChoresByArea: (area: AreaTag) => Chore[];
  getPendingChores: () => Chore[];
  getCompletedChores: () => Chore[];
  getOverdueChores: () => Chore[];
  getUpcomingChores: () => Chore[];
}

type ChoreStore = ChoreState & ChoreActions;

export const useChoreStore = create<ChoreStore>()(
  persist(
    (set, get) => ({
      chores: [],

      addChore: (chore) => {
        const id = uuidv4();
        const newChore: Chore = {
          ...chore,
          id,
          createdAt: Date.now(),
          isCompleted: false,
        };

        set((state) => ({
          chores: [newChore, ...state.chores],
        }));

        return id;
      },

      updateChore: (id, updates) => {
        set((state) => ({
          chores: state.chores.map((c) =>
            c.id === id ? { ...c, ...updates } : c
          ),
        }));
      },

      deleteChore: (id) => {
        set((state) => ({
          chores: state.chores.filter((c) => c.id !== id),
        }));
      },

      completeChore: (id) => {
        set((state) => ({
          chores: state.chores.map((c) =>
            c.id === id
              ? { ...c, isCompleted: true, completedAt: Date.now() }
              : c
          ),
        }));
      },

      uncompleteChore: (id) => {
        set((state) => ({
          chores: state.chores.map((c) =>
            c.id === id
              ? { ...c, isCompleted: false, completedAt: undefined }
              : c
          ),
        }));
      },

      addNote: (id, note) => {
        set((state) => ({
          chores: state.chores.map((c) =>
            c.id === id ? { ...c, notes: note } : c
          ),
        }));
      },

      getChoreById: (id) => {
        return get().chores.find((c) => c.id === id);
      },

      getChoresByArea: (area) => {
        return get().chores.filter((c) => c.areaTag === area && !c.isCompleted);
      },

      getPendingChores: () => {
        return get().chores.filter((c) => !c.isCompleted);
      },

      getCompletedChores: () => {
        return get().chores.filter((c) => c.isCompleted);
      },

      getOverdueChores: () => {
        const now = Date.now();
        return get().chores.filter(
          (c) => !c.isCompleted && c.dueDate && c.dueDate < now
        );
      },

      getUpcomingChores: () => {
        const now = Date.now();
        return get()
          .chores.filter((c) => !c.isCompleted && c.dueDate && c.dueDate >= now)
          .sort((a, b) => (a.dueDate || 0) - (b.dueDate || 0));
      },
    }),
    {
      name: "chore-storage",
      storage: createJSONStorage(() => AsyncStorage),
      partialize: (state) => ({
        chores: state.chores,
      }),
    }
  )
);
